### PR TITLE
Make it obvious how to create Control plugins that do not reference mapboxgl

### PR DESCRIPTION
**Pull request by tmcw**
## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs
 - [ ] post benchmark scores
 - [x] manually test the debug page

This implements part of #1835

### CHANGES

- Controls no longer take a `position` option. Instead, you provide a `position` argument to the `addControl` map method.
- Controls _must_ implement `onAdd` and `onRemove`.
- mapboxgl.Control: deleted
- Controls need to implement addTo -> controls need to implement onAdd
- attribution & scale controls are no longer evented

## Bugs fixed

- https://github.com/mapbox/mapbox-gl-js/issues/3561 no longer calls AttributionControl with `true` as an options argument
- https://github.com/mapbox/mapbox-gl-js/issues/3562 detaches event listeners from removed controls

cc @lucaswoj for the review

### CODE REVIEW

File: js/ui/control/attribution_control.js [hunk 1]
```diff
@@ -12,7 +12,7 @@ const DOM = require('../../util/dom');
  * var map = new mapboxgl.Map({attributionControl: false})
  *     .addControl(new mapboxgl.AttributionControl({position: 'top-left'}));
  */
-class AttributionControl extends Control {
+class AttributionControl extends Evented {
```
Comment (mourner) [inline js/ui/control/attribution_control.js]: `AttributionControl` doesn't need to be `Evented`.

File: js/ui/map.js [hunk 2]
```diff
@@ -212,14 +212,37 @@ class Map extends Camera {
...
+     * Adds a [`Control`](#Control) to the map, calling `control.onAdd(this)`.
      *
      * @param {Control} control The [`Control`](#Control) to add.
      * @returns {Map} `this`
      * @see [Display map navigation controls](https://www.mapbox.com/mapbox-gl-js/example/navigation/)
      */
     addControl(control) {
-        control.addTo(this);
+        control._map = this;
+        const container = control._container = control.onAdd(this);
```
Comment (lucaswoj) [inline js/ui/map.js]: We give the control a reference to the map via both the `_map` property and the `onAdd` method. Is there a reason to give this reference twice? Could we standardize on one?

File: js/ui/map.js [hunk 3]
```diff
@@ -212,14 +212,37 @@ class Map extends Camera {
...
+        if (control._position) {
+            const corner = this._controlCorners[control._position];
+            container.className += ' mapboxgl-ctrl';
+            if (control._position.indexOf('bottom') !== -1) {
```
Comment (lucaswoj) [inline js/ui/map.js]: If we're going to be accessing the `_position`, `_map`, and `_container` properties within `Map`, we should remove the `_underscore` to indicate that they are in the public API.

File: js/ui/map.js [hunk 4]
```diff
@@ -212,14 +212,37 @@ class Map extends Camera {
...
+                corner.insertBefore(container, corner.firstChild);
+            } else {
+                corner.appendChild(container);
+            }
+        }
+        return this;
```
Comment (lucaswoj) [inline js/ui/map.js]: Where / how do you plan to document the `Control` API?

**Comment by tmcw** [comment]
cc @lucaswoj for re-review.

**Comment by mourner** [comment]
@tmcw the build still fails.

**Comment by mourner** [comment]
```js
test/js/ui/map.test.js Map TypeError: Cannot use 'in' operator to search for 'nodeType' in undefined:
     Error: TypeError: Cannot use 'in' operator to search for 'nodeType' in undefined
      at Map.addControl (js/ui/map.js:78:488)
```

**Comment by tmcw** [comment]
Okay, fixed that test bug.

**Comment by tmcw** [comment]
ready for review @lucaswoj @mourner

**Comment by tmcw** [comment]
👍 switched from arrow functions to bindAll

File: js/ui/map.js [hunk 5]
```diff
@@ -212,14 +212,32 @@ class Map extends Camera {
...
+     * Adds a [`Control`](#Control) to the map, calling `control.onAdd(this)`.
      *
      * @param {Control} control The [`Control`](#Control) to add.
+     * @param {string} corner corner of the map to which the control will be added
      * @returns {Map} `this`
      * @see [Display map navigation controls](https://www.mapbox.com/mapbox-gl-js/example/navigation/)
      */
-    addControl(control) {
-        control.addTo(this);
+    addControl(control, corner = 'top-right') {
```
Comment (mourner) [inline js/ui/map.js]: Our default Node version for GL JS is still v4, so we can't use default parameters (which are v6+ I think).

File: js/ui/map.js [hunk 6]
```diff
@@ -1214,6 +1232,89 @@ function removeNode(node) {
...
+/**
+ * A [`LngLat`](#LngLat) object or an array of two numbers representing longitude and latitude.
+ *
+ * @typedef {(LngLat | Array<number>)} LngLatLike
+ * @example
+ * var v1 = new mapboxgl.LngLat(-122.420679, 37.772537);
+ * var v2 = [-122.420679, 37.772537];
+ */
+
+/**
```
Comment (mourner) [inline js/ui/map.js]: Looks like this got duplicated accidentally (same typedef below)

**Comment by mourner** [comment]
One last thing I'm thinking about — what do you think about readding the ability for the control to have a default position by adding an optional `getDefaultPosition()` method or `defaultPosition` property (optional) to the interface? This will make it slightly easier for people to use existing controls as they did before, without them having to update all `addControl` calls.

**Comment by tmcw** [comment]
👍 now implements getDefaultPosition, uses it for attributioncontrol and scalecontrol, and documents it in IControl

**Pull request merged**